### PR TITLE
ENG-8359 sync export overflow on snapshot

### PR DESCRIPTION
**Pull request by dweiss88**

### CODE REVIEW

File: src/frontend/org/voltdb/export/ExportManager.java [hunk 1]
```diff
@@ -692,4 +692,12 @@ public void truncateExportToTxnId(long snapshotTxnId, long[] perPartitionTxnIds)
             }
         }
     }
+
+    public static void syncExport(final boolean nofsync) {
```
Comment (akhanzode) [inline src/frontend/org/voltdb/export/ExportManager.java]: just call it sync()

File: src/frontend/org/voltdb/export/ExportGeneration.java [hunk 2]
```diff
@@ -784,6 +784,22 @@ public boolean truncateExportToTxnId(long txnId, long[] perPartitionTxnIds) {
         return m_drainedSources.get() == m_numSources;
     }
 
+    public void syncExport(final boolean nofsync) {
```
Comment (akhanzode) [inline src/frontend/org/voltdb/export/ExportGeneration.java]: just call it sync()

File: src/frontend/org/voltdb/export/ExportDataSource.java [hunk 3]
```diff
@@ -588,6 +588,28 @@ public void run() {
         }));
     }
 
+    private class SyncExportRunnable implements Runnable {
```
Comment (akhanzode) [inline src/frontend/org/voltdb/export/ExportDataSource.java]: Just call it sync()

File: src/frontend/org/voltdb/export/ExportGeneration.java [hunk 4]
```diff
@@ -784,6 +784,22 @@ public boolean truncateExportToTxnId(long txnId, long[] perPartitionTxnIds) {
...
+        List<ListenableFuture<?>> tasks = new ArrayList<ListenableFuture<?>>();
+
+        for (Map<String, ExportDataSource> dataSources : m_dataSourcesByPartition.values()) {
+            for (ExportDataSource source : dataSources.values()) {
+                tasks.add(source.syncExport(nofsync));
+            }
+        }
+
+        try {
+            Futures.allAsList(tasks).get();
```
Comment (akhanzode) [inline src/frontend/org/voltdb/export/ExportGeneration.java]: Call only if tasks is not empty

File: src/frontend/org/voltdb/export/ExportDataSource.java [hunk 5]
```diff
@@ -588,6 +588,28 @@ public void run() {
...
+        private final boolean nofsync;
+        SyncExportRunnable(final boolean nofsync) {
+            this.nofsync = nofsync;
+        }
+
+        @Override
+        public void run() {
+            try {
+                //Don't do a real sync, just write the in memory buffers
```
Comment (akhanzode) [inline src/frontend/org/voltdb/export/ExportDataSource.java]: fix comments to be more appropriate.

File: src/frontend/org/voltdb/export/ExportDataSource.java [hunk 6]
```diff
@@ -588,6 +588,28 @@ public void run() {
         }));
     }
 
+    private class SyncExportRunnable implements Runnable {
+        private final boolean nofsync;
```
Comment (akhanzode) [inline src/frontend/org/voltdb/export/ExportDataSource.java]: m_nofsync

File: src/frontend/org/voltdb/export/ExportDataSource.java [hunk 7]
```diff
@@ -588,6 +588,28 @@ public void run() {
...
+                //to a file. @Quiesce or blocking snapshot will do the sync
+                m_committedBuffers.sync(nofsync);
+            } catch (IOException e) {
+                exportLog.error(e);
+            }
+        }
+    }
+
+    public ListenableFuture<?> syncExport(final boolean nofsync) {
+        return m_es.submit(new SyncExportRunnable(nofsync));
```
Comment (akhanzode) [inline src/frontend/org/voltdb/export/ExportDataSource.java]: what if submit fails? catch the exception and propagate throwable.

File: src/frontend/org/voltdb/export/ExportDataSource.java [hunk 8]
```diff
@@ -588,6 +588,30 @@ public void run() {
...
+            this.m_nofsync = nofsync;
+        }
+
+        @Override
+        public void run() {
+            try {
+                m_committedBuffers.sync(m_nofsync);
+            } catch (IOException e) {
+                exportLog.error(e);
+            }
```
Comment (vtkstef) [inline src/frontend/org/voltdb/export/ExportDataSource.java]:
```java
exportLog.error("failed to sync export overflow",e);
```

File: src/frontend/org/voltdb/export/ExportDataSource.java [hunk 9]
```diff
@@ -588,6 +588,30 @@ public void run() {
...
+        }
+    }
+
+    public ListenableFuture<?> sync(final boolean nofsync) throws Exception {
```
Comment (vtkstef) [inline src/frontend/org/voltdb/export/ExportDataSource.java]: sure let's throw a really generic `Exception`, oh heck let's make it `Throwable`

create a separate runtime exception if you have to, and call it `ExportSyncException`,
and catch what submit can really throw and embed it as a cause to `ExportSyncException`

File: src/frontend/org/voltdb/export/ExportManager.java [hunk 10]
```diff
@@ -692,4 +692,12 @@ public void truncateExportToTxnId(long snapshotTxnId, long[] perPartitionTxnIds)
             }
         }
     }
+
+    public static void sync(final boolean nofsync) {
```
Comment (vtkstef) [inline src/frontend/org/voltdb/export/ExportManager.java]: synchronize it

File: src/frontend/org/voltdb/export/ExportDataSource.java [hunk 11]
```diff
@@ -588,6 +588,30 @@ public void run() {
...
+                exportLog.error("failed to sync export overflow", e);
+            }
+        }
+    }
+
+    public ListenableFuture<?> sync(final boolean nofsync) throws RejectedExecutionException {
+        try {
+            return m_es.submit(new SyncRunnable(nofsync));
+        } catch (RejectedExecutionException e) {
+            throw e;
```
Comment (vtkstef) [inline src/frontend/org/voltdb/export/ExportDataSource.java]: are you at least going to log it?!?!

Comment (dweiss88) [inline src/frontend/org/voltdb/export/ExportDataSource.java]: It get's logged by the calling function - ExportGeneration.sync

Comment (vtkstef) [inline src/frontend/org/voltdb/export/ExportDataSource.java]: so it implies that an inner method is aware of what the outer will do. Compartmentalize! Division of Concerns do they ring a bell for you?

**Pull request merged**